auth/role_info.cc

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "auth/role_info.hh"
+
+// Seastar features.
+#include <seastar/core/print.hh>
+
+// Scylla includes.
+#include "cql3/util.hh"
+#include "seastar/coroutine/maybe_yield.hh"
+
+namespace auth {
+
+std::ostream& role_info::describe(std::ostream& os) const {
+    // Note: fmt prints boolean values as `true` / `false` if the presentation type is not specified.
+
+    const auto formatted_role_name = cql3::util::maybe_quote(role_name);
+
+    if (maybe_salted_hash) {
+        // Note: Although `crypt` functions don't produce whitespace characters, etc., they don't guarantee
+        //       that the salted hash won't contain a quotation mark:
+        //
+        //         "Hashed passphrases are always entirely printable ASCII, and do not contain any whitespace
+        //          or the characters `:`, `;`, `*`, `!`, or `\`."
+        //
+        //       --- crypt(5); I changed the formatting of the manual page a bit.
+        //
+        //       Because of that, we call `cql3::util::maybe_quote` for safety.
+        //
+        // Note: `K_PASSWORD` in the grammar of CQL used by Scylla requires that passwords be quoted
+        //       with single quotation marks.
+        const auto salted_hash = cql3::util::single_quote(*maybe_salted_hash);
+
+        if (role_name == meta::DEFAULT_SUPERUSER_NAME) [[unlikely]] {
+            fmt::print(os, "CREATE ROLE IF NOT EXISTS {} WITH SALTED_HASH = {} AND LOGIN = {} AND SUPERUSER = {};",
+                    formatted_role_name, salted_hash, config.can_login, config.is_superuser);
+        } else {
+            fmt::print(os, "CREATE ROLE {} WITH SALTED_HASH = {} AND LOGIN = {} AND SUPERUSER = {};",
+                    formatted_role_name, salted_hash, config.can_login, config.is_superuser);
+        }
+    } else {
+        if (role_name == meta::DEFAULT_SUPERUSER_NAME) [[unlikely]] {
+            fmt::print(os, "CREATE ROLE IF NOT EXISTS {} WITH LOGIN = {} AND SUPERUSER = {};",
+                    formatted_role_name, config.can_login, config.is_superuser);
+        } else {
+            fmt::print(os, "CREATE ROLE {} WITH LOGIN = {} AND SUPERUSER = {};",
+                    formatted_role_name, config.can_login, config.is_superuser);
+        }
+    }
```

Comment:
Can we rework this code so that we don't have a separate branch for each combination of (is default role name, uses salted hash)?

For example:

```c++
sstring if_not_exists_part = ...;
sstring with_salted_hash_part = ...;

fmt::print(os, "CREATE ROLE{} {} WITH LOGIN = {} AND SUPERUSER = {}{}",
        if_not_exists_part, formatted_role_name, config.can_login, config.is_superuser, with_salted_hash_part);
```